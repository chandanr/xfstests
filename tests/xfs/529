#! /bin/bash
# SPDX-License-Identifier: GPL-2.0
# Copyright (c) 2020 Chandan Babu R.  All Rights Reserved.
#
# FS QA Test 529
#
# Verify that XFS does not cause inode fork's extent count to overflow when
# remapping extents from one file's inode fork to another.
seq=`basename $0`
seqres=$RESULT_DIR/$seq
echo "QA output created by $seq"

here=`pwd`
tmp=/tmp/$$
status=1	# failure is the default!
trap "_cleanup; exit \$status" 0 1 2 3 15

_cleanup()
{
	cd /
	rm -f $tmp.*
}

# get standard environment, filters and checks
. ./common/rc
. ./common/filter
. ./common/reflink
. ./common/inject

# remove previous $seqres.full before test
rm -f $seqres.full

# real QA test starts here

_supported_fs xfs
_require_scratch
_require_scratch_reflink
_require_xfs_debug
_require_xfs_io_command "reflink"
_require_xfs_io_error_injection "reduce_max_iextents"

echo "* Reflink remap extents"

echo "Format and mount fs"
_scratch_mkfs >> $seqres.full
_scratch_mount >> $seqres.full

bsize=$(_get_block_size $SCRATCH_MNT)

srcfile=${SCRATCH_MNT}/srcfile
dstfile=${SCRATCH_MNT}/dstfile

nr_blks=15

echo "Create \$srcfile having an extent of length $nr_blks blocks"
xfs_io -f -c "pwrite -b $((nr_blks * bsize))  0 $((nr_blks * bsize))" \
       -c fsync $srcfile >> $seqres.full

echo "Create \$dstfile having an extent of length $nr_blks blocks"
xfs_io -f -c "pwrite -b $((nr_blks * bsize))  0 $((nr_blks * bsize))" \
       -c fsync $dstfile >> $seqres.full

echo "Inject reduce_max_iextents error tag"
xfs_io -x -c 'inject reduce_max_iextents' $SCRATCH_MNT

echo "Reflink every other block from \$srcfile into \$dstfile"
for i in $(seq 1 2 $((nr_blks - 1))); do
	xfs_io -f -c "reflink $srcfile $((i * bsize)) $((i * bsize)) $bsize" \
	       $dstfile >> $seqres.full 2>&1
done

ino=$(stat -c "%i" $dstfile)

_scratch_unmount >> $seqres.full

echo "Verify \$dstfile's extent count"

nextents=$(_scratch_get_iext_count $ino data ||
		_fail "Unable to obtain inode fork's extent count")
if (( $nextents > 10 )); then
	echo "Extent count overflow check failed: nextents = $nextents"
fi

# success, all done
status=0
exit
