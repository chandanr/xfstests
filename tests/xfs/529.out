QA output created by 529
Format and mount fs
* Delalloc to written extent conversion
Inject reduce_max_iextents error tag
Create fragmented file
Verify $testfile's extent count
Disable reduce_max_iextents error tag
* Fallocate unwritten extents
Inject reduce_max_iextents error tag
Fallocate fragmented file
Verify $testfile's extent count
Disable reduce_max_iextents error tag
* Directio write
Inject reduce_max_iextents error tag
Create fragmented file via directio writes
Verify $testfile's extent count
Disable reduce_max_iextents error tag
* Extend quota inodes
Consume free space
Create fragmented filesystem
Inject reduce_max_iextents error tag
Inject bmap_alloc_minlen_extent error tag
Extend uquota file
Verify uquota inode's extent count
